File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums, k) {
        const freqMap = new Map();

        for (const num of nums) {
            freqMap.set(num, (freqMap.get(num) || 0) + 1);
        }

        const inverse = Array.from({ length: nums.length + 1 }, () => []);

        for (const [key, value] of freqMap) {
            inverse[value].push(key);
        }

        const highNumbers = [];

        for (let i = inverse.length - 1; i >= 0; i--) {
            if (inverse[i] && inverse[i].length) {
                highNumbers.push(...inverse[i]);

                if(highNumbers.length === k) break;
            }
        }

        return highNumbers;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = 1
        zero_count = 0

        for num in nums:
            if num == 0:
                zero_count = zero_count + 1
            else:
                product = product * num
        
        if zero_count > 1:
            return [0 for num in nums]
        
        if zero_count == 1:
            result = []
            for num in nums:
                if num == 0:
                    result.append(product)
                else:
                    result.append(0)
            
            return result
        
        result = []
        for num in nums:
            result.append(int(product / num))
        
        return result